From 191b59d699d21b3e154a15f6f345c90b457571ef Mon Sep 17 00:00:00 2001
From: Jan Pavlinec <jan.pavlinec@nic.cz>
Date: Fri, 28 May 2021 14:27:28 +0200
Subject: [PATCH] mvebu/patches: add fix for aardvark PCIe Completion
This fixes erratum 3.12 "PCIe completion timeout" related to mediatek AX cards
---
 ...lement-workaround-for-PCIe-Completio.patch | 25 +++++++++++++++++++
 1 file changed, 25 insertions(+)
 create mode 100644 target/linux/mvebu/patches-5.4/90126-PCI-aardvark-Implement-workaround-for-PCIe-Completio.patch

diff --git a/target/linux/mvebu/patches-5.4/90126-PCI-aardvark-Implement-workaround-for-PCIe-Completio.patch b/target/linux/mvebu/patches-5.4/90126-PCI-aardvark-Implement-workaround-for-PCIe-Completio.patch
new file mode 100644
index 0000000000..840d4f743b
--- /dev/null
+++ b/target/linux/mvebu/patches-5.4/90126-PCI-aardvark-Implement-workaround-for-PCIe-Completio.patch
@@ -0,0 +1,25 @@
+Index: linux-5.4.119/drivers/pci/controller/pci-aardvark.c
+===================================================================
+--- linux-5.4.119.orig/drivers/pci/controller/pci-aardvark.c
++++ linux-5.4.119/drivers/pci/controller/pci-aardvark.c
+@@ -133,6 +133,8 @@
+ #define     LTSSM_MASK				0x3f
+ #define     LTSSM_L0				0x10
+ #define     RC_BAR_CONFIG			0x300
++#define DEBUG_MUX_CTRL_REG			(LMI_BASE_ADDR + 0x208)
++#define     DIS_ORD_CHK				BIT(30)
+ 
+ /* PCIe core controller registers */
+ #define CTRL_CORE_BASE_ADDR			0x18000
+@@ -404,6 +406,11 @@ static void advk_pcie_setup_hw(struct ad
+ 	reg |= LANE_COUNT_1;
+ 	advk_writel(pcie, reg, PCIE_CORE_CTRL0_REG);
+ 
++	/* Disable ordering checks, workaround for erratum 3.12 "PCIe completion timeout" */
++	reg = advk_readl(pcie, DEBUG_MUX_CTRL_REG);
++	reg |= DIS_ORD_CHK;
++	advk_writel(pcie, reg, DEBUG_MUX_CTRL_REG);
++
+ 	/* Enable MSI */
+ 	reg = advk_readl(pcie, PCIE_CORE_CTRL2_REG);
+ 	reg |= PCIE_CORE_CTRL2_MSI_ENABLE;
-- 
2.25.1

